algebird-core/src/main/scala/com/twitter/algebird/macros/Cuber.scala

```diff
@@ -0,0 +1,81 @@
+package com.twitter.algebird.macros
+
+import scala.language.experimental.{ macros => sMacros }
+import scala.reflect.macros.Context
+import scala.reflect.runtime.universe._
+
+// import com.twitter.algebird._
+
+trait Cuber[I] {
+  type K
+  def apply(in: I): TraversableOnce[K]
+}
+
+trait Roller[I] {
+  type K
+  def apply(in: I): TraversableOnce[K]
+}
+
+object Cuber {
+  def cuber[T]: Cuber[T] = macro cuberImpl[T]
```

Comment:
the difference is that in scala, there is no optimizer that can avoid the allocation if you generate `Map[c.K, List[V]]` If you generate a TraversableOnce or Iterator, it is not so safe (you can't reuse it, and the type system does not know about ownership, consuming either does not have quite the same risk).

Also, scalding is different because we don't materialize the whole `TypedPipe/Grouped` into memory at once, so we can probably just have a `cube` method.

That said, are there good examples of things you want to do that are not semigroups in these cases (even commutative semigroups)? I'm not a big fan of giving people more general APIs than are commonly needed, when we have more constrained ones that have nice laws.